### PR TITLE
feat(bulk-import): allow adding repositories

### PR DESCRIPTION
**Pull request by debsmita1**
**Resolves:**
https://issues.redhat.com/browse/RHIDP-1664

**GIF:**

https://github.com/janus-idp/backstage-plugins/assets/22490998/e9f8e67d-63e3-426b-88ff-c6708c0917a7

### CODE REVIEW

**Comment by ShiranHi** [comment]
@debsmita1 it looks good! I am happy to see that! 🙌

I have few questions: 
1. Can we increase the width of the segmented button "Repositories" and "Organization"?
2.  If the status is "Failed generating file" we want to add a "Try again" button, and if users uncheck and checked again we'll try to generate it again ([Figma](https://www.figma.com/file/09Iltd57RiuqOtKWkjy8MH/Bulk-import?type=design&node-id=72%3A1254&mode=design&t=stWEmLJBJ9WuLCFs-1))
3. If ServiceNow selected the button label will be “Create ServiceNow tickets” instead of "Create pull requests"
4. btw, cam we change the label to be based on the number of repositores selected? if it is just one then the button label will be "Create pull request" (and not requests)? same question about "Create ServiceNow ticket"
5. I saw in the video that the title says "Selected repositories (7)" but there is one organization with 3 selected repositories and one with 2 selected repositories, should it be "Selected repositories (5)"?
6. When on the "Organization" view, I see one of them with the "Repository already added", I think we should not show that on the Repositories view unless all repositories from this organization added
7. If the repository already added the checkbox should be checked and disabled ([Figma](https://www.figma.com/file/09Iltd57RiuqOtKWkjy8MH/Bulk-import?type=design&node-id=11%3A4603&mode=design&t=gpY5PwhTjZYCJDVN-1))
8. Can we make the bottom bar with the buttons sticky to the bottom so users will always see it?

**Comment by debsmita1** [comment]
> 6. I saw in the video that the title says "Selected repositories (7)" but there is one organization with 3 selected repositories and one with 2 selected repositories, should it be "Selected repositories (5)"?

The tables have mock data as of now, As part of this PR I am only focusing on creating the tables and the components. But we should be able to handle it in this ticket https://issues.redhat.com/browse/RHIDP-1665

> 7. When on the "Organization" view, I see one of them with the "Repository already added", I think we should not show that on the Repositories view unless all repositories from this organization added

Should I leave it blank for cases where all the repositories are not added ? or Should I show "Not generated ?

> 10. Below the "Bulk import" side navigation item there is a border line, I think we don't need it

The divider was already present below  the Tech Radar. I simply grouped the Tech Radar and the Bulk Import nav items together

**Comment by ShiranHi** [comment]
> > 13. When on the "Organization" view, I see one of them with the "Repository already added", I think we should not show that on the Repositories view unless all repositories from this organization added
> 
> Should I leave it blank for cases where all the repositories are not added ? or Should I show "Not generated 
> 
If no respiratory from this org added then we will write "Not generated". If some added, let's write "x ready", if repositories from this org added, then we will write "Ready". What do you think?

> > 10. Below the "Bulk import" side navigation item there is a border line, I think we don't need it
> 
> The divider was already present below the Tech Radar. I simply grouped the Tech Radar and the Bulk Import nav items together
Got it, I think we don't need it

**Comment by debsmita1** [comment]
> > > 13. When on the "Organization" view, I see one of them with the "Repository already added", I think we should not show that on the Repositories view unless all repositories from this organization added
> > 
> > 
> > Should I leave it blank for cases where all the repositories are not added ? or Should I show "Not generated
> 
> If no respiratory from this org added then we will write "Not generated". If some added, let's write "x ready", if repositories from this org added, then we will write "Ready". What do you think?
> 

This makes sense

**Review by ciiay — COMMENTED** [review]
Can you confirm if `yarn start` works in the `plugins/bulk-import` directory? The previous version of this PR worked for me but this one doesn't.  After pulling upstream/main branch code and run `yarn install` `yarn start` it worked. The feature itself works good.

File: plugins/bulk-import/src/components/AddRepositories/RepositoriesTable.tsx
Comment (ciiay) [inline plugins/bulk-import/src/components/AddRepositories/RepositoriesTable.tsx:88]: Just a question, if user filtered data then click select all checkbox, will it select all data or only filtered data?

**Comment by invincibleJai** [comment]
@debsmita1 e2e tests are failing

File: plugins/bulk-import/src/components/AddRepositories/RepositoriesTable.tsx
Comment (debsmita1) [reply plugins/bulk-import/src/components/AddRepositories/RepositoriesTable.tsx]: Fixed it


https://github.com/janus-idp/backstage-plugins/assets/22490998/321443e5-960b-4839-bf0b-e8217cee63b4

**Comment by debsmita1** [comment]
> Can you confirm if `yarn start` works in the `plugins/bulk-import` directory? The previous version of this PR worked for me but this one doesn't. After pulling upstream/main branch code and run `yarn install` `yarn start` it worked. The feature itself works good.

Hey @ciiay ! It is working for me. Could you please try again ?

**Comment by ciiay** [comment]
@debsmita1 `yarn start` in `plugins/bulk-import/` now works for me too.

**Review by ciiay — APPROVED** [review]
LGTM
Thank you 👍

File: plugins/bulk-import/package.json
Comment (invincibleJai) [inline plugins/bulk-import/package.json:8]: i don't think this is needed, you might want to add https://github.com/janus-idp/backstage-plugins/blob/main/plugins/argocd/package.json#L7

File: plugins/bulk-import/src/components/AddRepositories/AddRepositoriesPage.tsx
Comment (invincibleJai) [inline plugins/bulk-import/src/components/AddRepositories/AddRepositoriesPage.tsx:59]: nit: can we avoid inline styles?

File: plugins/bulk-import/src/components/AddRepositories/AddRepositoriesTable.tsx
Comment (invincibleJai) [inline plugins/bulk-import/src/components/AddRepositories/AddRepositoriesTable.tsx:25]: nit: can we add types for this as well?

File: plugins/bulk-import/src/components/Repositories/RepositoriesList.tsx
Comment (invincibleJai) [inline plugins/bulk-import/src/components/Repositories/RepositoriesList.tsx:27]: should this `0` be always 0?

**Review by invincibleJai — COMMENTED** [review]
Thanks @debsmita1 , can we add unit tests as well as part of this PR?

**Comment by invincibleJai** [comment]
Here are some observations

- does CTA has to be sticky?  in dev mode it shows like below (and there are lot of empty space on right of first collapsible card)

image

- on clicking cancel on above CTA i see blank screen on dev mode

image

- Getting below warning

image

**Comment by ShiranHi** [comment]
@debsmita1 it looks good, that was fast 🙂

Few questions:
1. I got feedback on the top collapsable banner and we want to change the illustrations there ([Figma](https://www.figma.com/file/09Iltd57RiuqOtKWkjy8MH/Bulk-import?type=design&node-id=11%3A4603&mode=design&t=xXG6x9THlQgmk9Rm-1))
2. Let's try to horizontally align the icons there so they will fit the card width (as in the design)
3. We had a micro-copy review and the title of the card has changed to be "Add repositories to Red Hat Developer Hub in 5 steps" and the description text slightly changed too 
4. The bottom bar with the sticky buttons - can we make it higher? So we'll have 24px above and below it?
5. Tooltip appears on hover - can we prevent having a second line just for one word?
6. I saw on the first row that we write "Generating..." and there is "Try again" button next to it. I think we should not have scenario like this
7. If the status is "Failed generating file" and users check it again, we should re-try to create it, so after checking this repo, we should say "Generating..."

**Comment by debsmita1** [comment]
> Here are some observations
> 
> * does CTA has to be sticky?  in dev mode it shows like below (and there are lot of empty space on right of first collapsible card)
> 

Yes, the CTA has to be sticky. Fixed the empty space on the right

> image
> * on clicking cancel on above CTA i see blank screen on dev mode
> 

> image
> * Getting below warning
> 

Fixed it



> image

on it

**Comment by debsmita1** [comment]
> @debsmita1 it looks good, that was fast 🙂
> 
> Few questions:
> 
> 1. I got feedback on the top collapsable banner and we want to change the illustrations there ([Figma](https://www.figma.com/file/09Iltd57RiuqOtKWkjy8MH/Bulk-import?type=design&node-id=11%3A4603&mode=design&t=xXG6x9THlQgmk9Rm-1))
> 2. We had a micro-copy review and the title of the card has changed to be "Add repositories to Red Hat Developer Hub in 5 steps" and the description text slightly changed too
> 3. The bottom bar with the sticky buttons - can we make it higher? So we'll have 24px above and below it?
> 4. Tooltip appears on hover - can we prevent having a second line just for one word?
> 5. I saw on the first row that we write "Generating..." and there is "Try again" button next to it. I think we should not have scenario like this
> 6. If the status is "Failed generating file" and users check it again, we should re-try to create it, so after checking this repo, we should say "Generating..."
> 7. Now that I see it, the "Approval tool" is quite disappearing, do you mind moving them to the top of the page? I hope we have time for this improvement ([Figma](https://www.figma.com/file/09Iltd57RiuqOtKWkjy8MH/Bulk-import?type=design&node-id=392%3A599&mode=design&t=xXG6x9THlQgmk9Rm-1))

Updated the GIF. PTAL

File: plugins/bulk-import/src/components/AddRepositories/AddRepositoriesPage.tsx
Comment (ciiay) [inline plugins/bulk-import/src/components/AddRepositories/AddRepositoriesPage.tsx:36]: We can use `display: flex` in the Illustrations container and use `justify-content: space-around` instead of hard coded right padding of illustrations.
rhidp_1664_pr_review

Comment (debsmita1) [reply plugins/bulk-import/src/components/AddRepositories/AddRepositoriesPage.tsx:36]: Adding `display: flex` to the container wrapping the illustrations, the mobile view looks like this
Screenshot 2024-04-05 at 10 41 06 AM

**Comment by ShiranHi** [comment]
@debsmita1 thank you for the last minute changes, it looks much better now 🙌

Two questions:
1. The line below the "Approval tool" seems too bright, can we make sure it is the same as in the design ([Figma](https://www.figma.com/file/09Iltd57RiuqOtKWkjy8MH/Bulk-import?type=design&node-id=392%3A599&mode=design&t=GFVjQ7qpvPeDYLAz-1))?
2. Sticky buttons - do they have a transparent background? Cab we make sure it is white background and has a line above it?

File: plugins/bulk-import/src/components/AddRepositories/AddRepositoriesPage.tsx
Comment (ciiay) [reply plugins/bulk-import/src/components/AddRepositories/AddRepositoriesPage.tsx:36]: Oh I didn't know we need to deal with mobile view as well. In that case is adding a `overflow: scroll` acceptable? The current padding doesn't work on my laptop default screen. There might be other displays we need to review and get some layout design from UX team for better mobile screen display.

https://github.com/janus-idp/backstage-plugins/assets/26255262/0ef6706e-a585-4321-b51a-7a17b8d689da

**Comment by invincibleJai** [comment]
- In dev mode i see below warnings, check if this is relevant

image

image

- on the right in the collapsible panel I see it has more empty space and even figma has some @ShiranHi is this expected?

image

Figma:

image

**Comment by debsmita1** [comment]
> @debsmita1 thank you for the last minute changes, it looks much better now 🙌
> 
> Two questions:
> 
> 1. The line below the "Approval tool" seems too bright, can we make sure it is the same as in the design ([Figma](https://www.figma.com/file/09Iltd57RiuqOtKWkjy8MH/Bulk-import?type=design&node-id=392%3A599&mode=design&t=GFVjQ7qpvPeDYLAz-1))?
> 2. Sticky buttons - do they have a transparent background? Cab we make sure it is white background and has a line above it?

@ShiranHi  the background is not transparent

Screenshot 2024-04-08 at 8 10 09 PM

**Review by ciiay — COMMENTED** [review]
The According works great on my end now. Thanks for the update.
There appears a line above bottom buttons which wasn't there before. Is this added on purpose?

**Comment by ShiranHi** [comment]
> * on the right in the collapsible panel I see it has more empty space and even figma has some @ShiranHi is this expected?
Yes, that is ok. It will be great if we can have some padding on the left side, just before the first icon. 

>The line below the "Approval tool" seems too bright, can we make sure it is the same as in the design ([Figma](https://www.figma.com/file/09Iltd57RiuqOtKWkjy8MH/Bulk-import?type=design&node-id=392%3A599&mode=design&t=GFVjQ7qpvPeDYLAz-1))?
Thank you for fixing this, now it seems too dark, can we make sure it is has the same gray color as the segmented button border (with the Repository and the Organization buttons)?

>@ShiranHi the background is not transparent
Can we have a White background there and that the line will not have padding on the right side?
image

**Comment by ciiay** [comment]
>Can we have a White background there and that the line will not have padding on the right side?

Hi @ShiranHi @debsmita1 The white background setting will be addressed as default main container background color in [this PR](https://github.com/janus-idp/backstage-showcase/pull/1171)  on backstage-showcase.

**Comment by debsmita1** [comment]
> > * on the right in the collapsible panel I see it has more empty space and even figma has some @ShiranHi is this expected?
> 
> Yes, that is ok. It will be great if we can have some padding on the left side, just before the first icon.
> 
> > The line below the "Approval tool" seems too bright, can we make sure it is the same as in the design ([Figma](https://www.figma.com/file/09Iltd57RiuqOtKWkjy8MH/Bulk-import?type=design&node-id=392%3A599&mode=design&t=GFVjQ7qpvPeDYLAz-1))?
> 
> Thank you for fixing this, now it seems too dark, can we make sure it is has the same gray color as the segmented button border (with the Repository and the Organization buttons)?
> 
> > @ShiranHi the background is not transparent
> 
> Can we have a White background there and that the line will not have padding on the right side? image

@ShiranHi Updated , PTAL
Screenshot 2024-04-11 at 9 23 51 AM
 
 @ciiay ^

**Comment by ShiranHi** [comment]
@debsmita1 looks good to me, thank you!

**Comment by invincibleJai** [comment]
in dark theme this looks like this

image

@debsmita1 is this expected?

cc @ShiranHi

**Comment by ShiranHi** [comment]
> in dark theme this looks like this
> 
> image

Right, good catch! We need to use the white version for dark theme ([icons are here](https://redhat.brand-portal.adobe.com/mediaportal.html/content/dam/mac/redhat/brand/illustrations/mini-illustrations))

File: plugins/bulk-import/src/components/AddRepositories/AddRepositoriesForm.tsx
Comment (debsmita1) [reply plugins/bulk-import/src/components/AddRepositories/AddRepositoriesForm.tsx:65]: yes @invincibleJai 
Screenshot 2024-04-11 at 3 06 20 PM

@ShiranHi  ^

Comment (ShiranHi) [reply plugins/bulk-import/src/components/AddRepositories/AddRepositoriesForm.tsx:65]: > this text is as per design? got confused with
> 
> `Once it's approved and the PR is approved`

oops, that is a mistake, this is the text:
When adding a new repository, it requires approval. Once the PR is approved or the ServiceNow ticket is closed, the repositories will be added to the Catalog page.

**Comment by ShiranHi** [comment]
> image

Remove the "Please" word at the beginning, thanks!

**Comment by debsmita1** [comment]
> in dark theme this looks like this
> 
> image
> and click on repo / org open's in new window, which is expected should we have `launch` icon as we have designs image
> 
> image
> @debsmita1 ^^
> 
> cc @ShiranHi



Done

Screenshot 2024-04-11 at 3 42 59 PM

@invincibleJai @ShiranHi

File: plugins/bulk-import/src/components/AddRepositories/AddRepositoriesForm.tsx
Comment (debsmita1) [reply plugins/bulk-import/src/components/AddRepositories/AddRepositoriesForm.tsx]: Done

Screenshot 2024-04-11 at 3 42 59 PM

**Pull request merged**
